src/tdlib/client/mod.rs

```diff
@@ -0,0 +1,263 @@
+};
+
+/// A struct for storing information about a session's database.
+#[derive(Clone, Debug)]
+pub(crate) struct DatabaseInfo {
+    // The base name of the database directory.
+    pub(crate) directory_base_name: String,
+    // Whether this database uses a test dc.
+    pub(crate) use_test_dc: bool,
+}
+
+mod imp {
+    use super::*;
+
+    #[derive(Debug, Default, Properties)]
+    #[properties(wrapper_type = super::Client)]
+    pub(crate) struct Client {
+        #[property(get, set, construct_only)]
+        pub(super) manager: WeakRef<ClientManager>,
+        #[property(get, set, construct_only)]
+        pub(super) id: Cell<i32>,
+        #[property(get, set, construct_only)]
+        pub(super) database_info: OnceCell<BoxedDatabaseInfo>,
+        #[property(get)]
+        pub(super) state: RefCell<Option<glib::Object>>,
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for Client {
+        const NAME: &'static str = "Client";
+        type Type = super::Client;
+    }
+
+    impl ObjectImpl for Client {
+        fn properties() -> &'static [glib::ParamSpec] {
+            Self::derived_properties()
+        }
+        fn set_property(&self, id: usize, value: &glib::Value, pspec: &glib::ParamSpec) {
+            self.derived_set_property(id, value, pspec)
+        }
+        fn property(&self, id: usize, pspec: &glib::ParamSpec) -> glib::Value {
+            self.derived_property(id, pspec)
+        }
+
+        fn constructed(&self) {
+            self.parent_constructed();
+
+            let obj = &*self.obj();
+
+            spawn(clone!(@weak obj => async move {
+                obj.init().await;
+            }));
+            obj.set_state(ClientAuth::from(obj).upcast());
+        }
+    }
+}
+
+glib::wrapper! { pub(crate) struct Client(ObjectSubclass<imp::Client>); }
+
+impl Client {
+    pub(crate) fn new(
+        manager: &ClientManager,
+        client_id: i32,
+        database_info: DatabaseInfo,
+    ) -> Self {
+        glib::Object::builder()
+            .property("manager", manager)
+            .property("id", client_id)
+            .property("database-info", BoxedDatabaseInfo(database_info))
+            .build()
+    }
+
+    async fn init(&self) {
+        let result = functions::set_log_verbosity_level(
+            if log::log_enabled!(log::Level::Trace) {
+                5
+            } else if log::log_enabled!(log::Level::Debug) {
+                4
+            } else if log::log_enabled!(log::Level::Info) {
+                3
+            } else if log::log_enabled!(log::Level::Warn) {
+                2
+            } else {
+                0
+            },
+            self.id(),
+        )
+        .await;
+
+        if let Err(e) = result {
+            log::warn!("Error setting the tdlib log level: {:?}", e);
+        }
+
+        // TODO: Hopefully we'll support animated emoji at some point
+        let result = self.disable_animated_emoji(true).await;
```

Comment:
```suggestion
        let result = self.disable_animated_emoji(false).await;
```

I guess you worked on this PR before animated emojis were supported.

Reply:
Need to change the warn message below too.

Reply:
Yes, this is probably one of the many things I broke during rebase. In the end, I will visit every change carefully.